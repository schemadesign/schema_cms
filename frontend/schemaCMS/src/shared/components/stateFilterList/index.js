export { StateFilterList } from './stateFilterList.component';
