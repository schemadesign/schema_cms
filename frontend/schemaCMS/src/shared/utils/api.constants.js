const API_PREFIX = process.env.REACT_APP_BASE_API_URL;

export const AUTH_PATH = `${API_PREFIX}/auth/login/auth0`;
export const TOKEN_PATH = '/auth/token/';
export const USER_PATH = '/user/';
