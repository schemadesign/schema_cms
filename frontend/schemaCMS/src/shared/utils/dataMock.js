export const tableData = [
  {
    'First Name': 'George',
    'Last Name': 'Washington',
    'Full Name': 'George Washington',
    Birthplace: 'Westmoreland County',
    'State of Birth': 'VA',
    Birthday: '02/22/1732',
    Url: 'some.url.com',
  },
  {
    'First Name': 'John',
    'Last Name': 'Adams',
    'Full Name': 'John Adams',
    Birthplace: 'Braintree',
    'State of Birth': 'MA',
    Birthday: '10/30/1735',
    Url: 'some.url.com',
  },
  {
    'First Name': 'Thomas',
    'Last Name': 'Jefferson',
    'Full Name': 'Thomas Jefferson',
    Birthplace: 'Shadwell',
    'State of Birth': 'VA',
    Birthday: '04/13/1743',
    Url: 'some.url.com',
  },
  {
    'First Name': 'James',
    'Last Name': 'Madison',
    'Full Name': 'James Madison',
    Birthplace: 'Port Conway',
    'State of Birth': 'VA',
    Birthday: '03/16/1751',
    Url: 'some.url.com',
  },
  {
    'First Name': 'James',
    'Last Name': 'Monroe',
    'Full Name': 'James Monroe',
    Birthplace: 'Monroe Hall',
    'State of Birth': 'VA',
    Birthday: '04/28/1758',
    Url: 'some.url.com',
  },
];

export const tableFields = {
  'First Name': {},
  'Last Name': {},
  'Full Name': {},
  Birthplace: {},
  'State of Birth': {},
  Birthday: {},
  Url: {},
};
