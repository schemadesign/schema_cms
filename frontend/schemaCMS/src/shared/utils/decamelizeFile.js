import { decamelize } from 'humps';

export const decamelizeThatDontBreaksFile = object => {
  if (object && !(object instanceof File)) {
    if (object instanceof Array) {
      return object.map(item => decamelizeThatDontBreaksFile(item));
    }
    if (object instanceof FormData) {
      let formData = new FormData();
      for (const [key, value] of object.entries()) {
        formData.append(decamelize(key), value);
      }
      return formData;
    }
    if (typeof object === 'object') {
      return Object.keys(object).reduce(
        (acc, next) => ({
          ...acc,
          [decamelize(next)]: decamelizeThatDontBreaksFile(object[next]),
        }),
        {}
      );
    }
  }
  return object;
};
