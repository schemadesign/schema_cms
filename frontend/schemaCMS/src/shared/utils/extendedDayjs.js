import dayjs from 'dayjs';
import relativeTime from 'dayjs/plugin/relativeTime';
import customParseFormat from 'dayjs/plugin/customParseFormat';

dayjs.extend(relativeTime).extend(customParseFormat);

export default dayjs;

export const BASE_DATE_FORMAT = 'YYYY-MM-DDTHH:mm:ss.000ZZ';
