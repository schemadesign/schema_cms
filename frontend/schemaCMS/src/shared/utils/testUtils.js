import { Router } from 'react-router';
import { IntlProvider } from 'react-intl';
import { act, create } from 'react-test-renderer';
import React from 'react';
import { ThemeProvider } from 'styled-components';
import { Theme } from 'schemaUI';
import PropTypes from 'prop-types';

import { DEFAULT_LOCALE, translationMessages } from '../../i18n';
import browserHistory from './history';

const ProvidersWrapper = ({ children }) => (
  <Router history={browserHistory}>
    <IntlProvider locale={DEFAULT_LOCALE} messages={translationMessages[DEFAULT_LOCALE]}>
      <ThemeProvider theme={Theme.dark}>{children}</ThemeProvider>
    </IntlProvider>
  </Router>
);

ProvidersWrapper.propTypes = {
  children: PropTypes.node.isRequired,
};

export const makeContextRenderer = async component => {
  let wrapper = null;

  await act(async () => {
    wrapper = await create(<ProvidersWrapper>{component}</ProvidersWrapper>);
  });

  return wrapper;
};
