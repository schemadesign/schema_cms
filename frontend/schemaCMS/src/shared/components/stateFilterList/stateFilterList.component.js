import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { Form as FormUI } from 'schemaUI';
import { always, append, equals, ifElse, reject } from 'ramda';
import { useHistory, useLocation } from 'react-router';

import messages from './stateFilterList.messages';
import { LoadingWrapper } from '../loadingWrapper';
import {
  DATA_SOURCE_STATE_ACTIVE_FILTERS,
  DATA_SOURCE_STATE_FILTERS,
} from '../../../modules/dataSourceState/dataSourceState.constants';
import { FilterName, containerCheckboxGroupStyles } from './stateFilterList.styles';

const { CheckboxGroup, Checkbox, Label } = FormUI;

export const StateFilterList = ({ filters, values, setFieldValue }) => {
  const history = useHistory();
  const { pathname } = useLocation();
  const handleGoToFilter = id => history.push(`/state/filter/${id}`, { backUrl: pathname, state: values });
  const handleChange = e => {
    const { value, checked } = e.target;
    const intValue = parseInt(value, 10);
    const isFilled = values.filters.find(({ filter }) => filter === intValue);

    if (checked && !isFilled) {
      return handleGoToFilter(value);
    }

    const setFilters = ifElse(
      equals(true),
      always(append(intValue, values[DATA_SOURCE_STATE_ACTIVE_FILTERS])),
      always(reject(equals(intValue), values[DATA_SOURCE_STATE_ACTIVE_FILTERS]))
    );

    return setFieldValue(DATA_SOURCE_STATE_ACTIVE_FILTERS, setFilters(checked));
  };

  return (
    <Fragment>
      <Label>
        <FormattedMessage {...messages[DATA_SOURCE_STATE_FILTERS]} />
      </Label>
      <LoadingWrapper noData={!filters.length} noDataContent={<FormattedMessage {...messages.noData} />}>
        <CheckboxGroup
          onChange={handleChange}
          name={DATA_SOURCE_STATE_FILTERS}
          value={values[DATA_SOURCE_STATE_ACTIVE_FILTERS]}
          customStyles={containerCheckboxGroupStyles}
        >
          {filters.map(({ id, name }, index) => (
            <Checkbox key={index} id={`checkbox-${index}`} value={id}>
              <FilterName onClick={() => handleGoToFilter(id)}>{name}</FilterName>
            </Checkbox>
          ))}
        </CheckboxGroup>
      </LoadingWrapper>
    </Fragment>
  );
};

StateFilterList.propTypes = {
  state: PropTypes.object.isRequired,
  values: PropTypes.object.isRequired,
  filters: PropTypes.array.isRequired,
  setFieldValue: PropTypes.func.isRequired,
};
