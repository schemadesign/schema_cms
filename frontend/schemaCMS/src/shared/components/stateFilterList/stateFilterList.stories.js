import React from 'react';
import { storiesOf } from '@storybook/react';

import { StateFilterList } from './stateFilterList.component';
import { state } from '../../../modules/dataSourceState/dataSourceState.mock';
import { withTheme } from '../../../.storybook/decorators';
import { filters } from '../../../modules/filter/filter.mocks';

export const defaultProps = {
  setFieldValue: Function.prototype,
  state,
  filters,
  values: { filters: [{ filter: 1, values: ['George'] }, { filter: 2, values: ['Washington'] }], activeFilters: [1] },
};

storiesOf('Shared Components|StateFilterList', module)
  .addDecorator(withTheme())
  .add('Default', () => <StateFilterList {...defaultProps} />);
