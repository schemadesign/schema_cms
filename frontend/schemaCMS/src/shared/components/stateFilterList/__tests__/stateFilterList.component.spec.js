import React from 'react';
import { act } from 'react-test-renderer';
import { Form } from 'schemaUI';

import { StateFilterList } from '../stateFilterList.component';
import { defaultProps } from '../stateFilterList.stories';
import { makeContextRenderer } from '../../../utils/testUtils';

const { CheckboxGroup } = Form;
const mockPushHistory = jest.fn();

jest.mock('react-router', () => ({
  ...jest.requireActual('react-router'),
  useHistory: () => ({
    push: mockPushHistory,
  }),
}));

describe('StateFilterList: Component', () => {
  const render = props => makeContextRenderer(<StateFilterList {...defaultProps} {...props} />);

  it('should render correctly', async () => {
    const wrapper = await render();
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should deselect filter', async () => {
    jest.spyOn(defaultProps, 'setFieldValue');
    const wrapper = await render();

    act(() => {
      wrapper.root.findByType(CheckboxGroup).props.onChange({ target: { value: '1', checked: false } });
    });

    await Promise.resolve();
    expect(defaultProps.setFieldValue).toHaveBeenCalledWith('activeFilters', []);
  });

  it('should select filter', async () => {
    jest.spyOn(defaultProps, 'setFieldValue');
    const wrapper = await render();

    act(() => {
      wrapper.root.findByType(CheckboxGroup).props.onChange({ target: { value: '2', checked: true } });
    });

    await Promise.resolve();
    expect(defaultProps.setFieldValue).toHaveBeenCalledWith('activeFilters', [1, 2]);
  });

  it('should redirect to filter if filter is not set', async () => {
    jest.spyOn(defaultProps, 'setFieldValue');
    const wrapper = await render();

    act(() => {
      wrapper.root.findByType(CheckboxGroup).props.onChange({ target: { value: '3', checked: true } });
    });

    await Promise.resolve();
    expect(mockPushHistory).toHaveBeenCalledWith('/state/filter/3', {
      backUrl: '/',
      state: {
        activeFilters: [1],
        filters: [{ filter: 1, values: ['George'] }, { filter: 2, values: ['Washington'] }],
      },
    });
  });

  it('should redirect to filter', async () => {
    const wrapper = await render();

    act(() => {
      wrapper.root.findByProps({ id: 'filterName-1' }).props.onClick();
    });

    await Promise.resolve();
    expect(mockPushHistory).toHaveBeenCalledWith('/state/filter/2', {
      backUrl: '/',
      state: {
        activeFilters: [1],
        filters: [{ filter: 1, values: ['George'] }, { filter: 2, values: ['Washington'] }],
      },
    });
  });
});
