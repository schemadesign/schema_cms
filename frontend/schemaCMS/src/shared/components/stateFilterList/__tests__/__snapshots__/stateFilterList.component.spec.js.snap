// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateFilterList: Component should render correctly 1`] = `
Array [
  <label
    style={
      Object {
        "border": "none",
        "borderTop": "2px solid #2F2F36",
        "color": "#71737e",
        "display": "block",
        "fontSize": "14px",
        "outline": "none",
        "paddingTop": "8px",
        "width": "100%",
      }
    }
  >
    Filters
  </label>,
  <div
    style={
      Object {
        "border": "none",
        "borderTop": "1px solid #2F2F36",
        "display": "flex",
        "flexDirection": "column",
        "padding": "20px 20px 40px",
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "borderBottom": "1px solid #2F2F36",
          "color": "#FFF",
          "display": "flex",
          "flexDirection": "row",
          "height": 48,
          "justifyContent": "space-between",
        }
      }
    >
      <input
        aria-hidden={true}
        checked={true}
        id="checkbox-0"
        name="filters"
        onChange={[Function]}
        style={
          Object {
            "height": 0,
            "left": 0,
            "overflow": "hidden",
            "position": "absolute",
            "top": 0,
            "visibility": "hidden",
          }
        }
        type="checkbox"
        value={1}
      />
      <span
        style={
          Object {
            "overflow": "hidden",
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <div
          className="sc-AxhCb iniznJ"
          id="filterName-0"
          onClick={[Function]}
        >
          Filter First Name
        </div>
      </span>
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "fontSize": 0,
          }
        }
      >
        <label
          htmlFor="checkbox-0"
          style={
            Object {
              "cursor": "pointer",
              "marginRight": 0,
            }
          }
        >
          <svg
            stroke="#000"
            style={
              Object {
                "height": "30px",
                "stroke": "#FFF",
                "width": "30px",
              }
            }
            viewBox="0 0 30 30"
          >
            <g
              fill="none"
              fillRule="evenodd"
            >
              <path
                d="M20.455 7.5H7.5v15h15v-6.136"
                strokeLinecap="square"
                strokeLinejoin="round"
              />
              <path
                d="M11 12l5 5L26 7"
                strokeWidth={3}
              />
            </g>
          </svg>
        </label>
      </div>
    </div>
    <div
      style={
        Object {
          "alignItems": "center",
          "borderBottom": "1px solid #2F2F36",
          "color": "#FFF",
          "display": "flex",
          "flexDirection": "row",
          "height": 48,
          "justifyContent": "space-between",
        }
      }
    >
      <input
        aria-hidden={true}
        checked={false}
        id="checkbox-1"
        name="filters"
        onChange={[Function]}
        style={
          Object {
            "height": 0,
            "left": 0,
            "overflow": "hidden",
            "position": "absolute",
            "top": 0,
            "visibility": "hidden",
          }
        }
        type="checkbox"
        value={2}
      />
      <span
        style={
          Object {
            "overflow": "hidden",
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <div
          className="sc-AxhCb iniznJ"
          id="filterName-1"
          onClick={[Function]}
        >
          Filter Last Name
        </div>
      </span>
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "fontSize": 0,
          }
        }
      >
        <label
          htmlFor="checkbox-1"
          style={
            Object {
              "cursor": "pointer",
              "marginRight": 0,
            }
          }
        >
          <svg
            style={
              Object {
                "height": "30px",
                "stroke": "#FFF",
                "width": "30px",
              }
            }
            viewBox="0 0 30 30"
          >
            <path
              d="M7.5 7.5h15v15h-15z"
              fill="none"
            />
          </svg>
        </label>
      </div>
    </div>
    <div
      style={
        Object {
          "alignItems": "center",
          "borderBottom": "1px solid #2F2F36",
          "color": "#FFF",
          "display": "flex",
          "flexDirection": "row",
          "height": 48,
          "justifyContent": "space-between",
        }
      }
    >
      <input
        aria-hidden={true}
        checked={false}
        id="checkbox-2"
        name="filters"
        onChange={[Function]}
        style={
          Object {
            "height": 0,
            "left": 0,
            "overflow": "hidden",
            "position": "absolute",
            "top": 0,
            "visibility": "hidden",
          }
        }
        type="checkbox"
        value={3}
      />
      <span
        style={
          Object {
            "overflow": "hidden",
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <div
          className="sc-AxhCb iniznJ"
          id="filterName-2"
          onClick={[Function]}
        >
          Filter Birthday
        </div>
      </span>
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "fontSize": 0,
          }
        }
      >
        <label
          htmlFor="checkbox-2"
          style={
            Object {
              "cursor": "pointer",
              "marginRight": 0,
            }
          }
        >
          <svg
            style={
              Object {
                "height": "30px",
                "stroke": "#FFF",
                "width": "30px",
              }
            }
            viewBox="0 0 30 30"
          >
            <path
              d="M7.5 7.5h15v15h-15z"
              fill="none"
            />
          </svg>
        </label>
      </div>
    </div>
  </div>,
]
`;
